src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorSourceUpdater.cs

```diff
@@ -48,59 +56,332 @@ public async Task<IUpdaterResult> IsApplicableAsync(IUpgradeContext context, Imm
+            var diagnosticsByFile = mappedDiagnostics.GroupBy(d => d.Location.GetMappedLineSpan().Path);
+            foreach (var diagnosticsGroup in diagnosticsByFile)
+            {
+                _logger.LogInformation("  {DiagnosticsCount} diagnostics need fixed in {FilePath}", diagnosticsGroup.Count(), diagnosticsGroup.Key);
+            }
+
+            return new FileUpdaterResult(mappedDiagnostics.Any(), diagnosticsByFile.Select(g => g.Key));
+        }
+
+        public async Task<IUpdaterResult> ApplyAsync(IUpgradeContext context, ImmutableArray<RazorCodeDocument> inputs, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var project = GetProjectWithGeneratedCode(context.CurrentProject.Required(), inputs);
+            var originalProject = project;
+            var generatedFilePaths = inputs.Select(GetGeneratedFilePath);
+
+            // Regenerating diagnostics is slow for large projects, but is necessary in between fixing multiple diagnostics
+            // in a single file. To minimize the number of time diagnostics are gathered, fix one diagnostic each
+            // from multiple files before regenerating diagnostics.
+            var diagnostics = await GetDiagnosticsFromTargetFilesAsync(project, context, generatedFilePaths, new LocationAndIDComparer(false), token).ConfigureAwait(false);
+            var diagnosticsCount = diagnostics.Count();
+            while (diagnosticsCount > 0)
+            {
+                // Iterate through the first diagnostic from each document
+                foreach (var diagnostic in diagnostics.GroupBy(d => d.Location.SourceTree?.FilePath).Select(g => g.First()))
+                {
+                    var doc = project.GetDocument(diagnostic.Location.SourceTree);
+                    if (doc is null)
+                    {
+                        continue;
+                    }
+
+                    // Apply the code fix in the generated source
+                    project = await TryFixDiagnosticAsync(diagnostic, doc).ConfigureAwait(false);
+                }
+
+                diagnostics = await GetDiagnosticsFromTargetFilesAsync(project, context, generatedFilePaths, new LocationAndIDComparer(false), token).ConfigureAwait(false);
+                var newDiagnosticCount = diagnostics.Count();
+                if (diagnosticsCount == newDiagnosticCount)
+                {
+                    _logger.LogWarning("Diagnostics could not be fixed as expected. This may be caused by the project being in a bad state (did NuGet packages restore correctly?) or by errors in analyzers or code fix providers related to diagnosticsL {DiagnosticIds}.", string.Join(", ", diagnostics.Select(d => d.Id).Distinct()));
```

Comment:
is there a call to action here? seems like we want to ask the user to confirm that their packages were restored and maybe other things.

Reply:
Not really, it's just a generic warning that the diagnostics weren't fixed as expected so something went wrong. It's hard to know what, though. Could be that packages weren't restored but could also be a bug in the analyzer or code fix provider.